src/services/BaseApi.ts

```diff
@@ -0,0 +1,128 @@
+import axios, { AxiosInstance, AxiosRequestConfig, AxiosRequestHeaders, AxiosResponse } from 'axios'
+
+export type PrefectConfig = {
+  baseUrl: string,
+  token?: string,
+}
+
+export type GetApiBaseUrl<T extends PrefectConfig = PrefectConfig> = (config: T) => string
+export type GetApiHeaders<T extends PrefectConfig = PrefectConfig> = (config: T) => AxiosRequestHeaders
+
+export const getPrefectBaseUrl: GetApiBaseUrl = (config) => {
+  return config.baseUrl
+}
+
+export const getPrefectUIHeaders: GetApiHeaders = () => {
+  return {
+    'X-PREFECT-UI': true,
+  }
+}
+
+export const getAuthorizationHeaders: GetApiHeaders = (config) => {
+  const value: AxiosRequestHeaders = {}
+
+  if (config.token) {
+    value.Authorization = `bearer ${config.token}`
+  }
+
+  return value
+}
+
+export class BaseApi<T extends PrefectConfig = PrefectConfig> {
+  protected readonly apiConfig: T
+  protected routePrefix: string | undefined
+  protected getHeaders: GetApiHeaders[]
+  protected getBaseUrl: GetApiBaseUrl
```

Comment:
I feel like we could make these more flexible with very little effort by updating the types a bit

```typescript
type ApiHeaders = MaybeArray<AxiosRequestHeaders | GetApiHeaders>
type ApiBaseUrl = string | GetApiBaseUrl
```

Reply:
I also think these should drop the `get` prefix and be just `headers` and `baseUrl`. Then rename the `compose` methods to `getHeaders` and `getBaseUrl`. IMO a little more clear as to what its doing.

Reply:
I don't feel strongly about either comment above. More strongly about the naming than the types.

Reply:
I agree with both comments